### PR TITLE
Make base expiry configurable

### PR DESCRIPTION
**Pull request by mmartinson**
Here's a first pass at making the expiry configurable as I was asking in #17. Let me know if you think this works. I added a bit more to the README as well about how you might shoot yourself in the foot with a shorter expiry.

One small uncertainty I had while working on this was wondering if there is any reason for the RESERVED expiry to be different from the IN_PROGRESS attempt-1 expiry. The RESERVED expiry seems like it would rare, at least running on a single node. Could you see a reason that this maybe remain a 5-minute constant, where the IN_PROGRESS should be configurable. Or should they be configurable separately?

### CODE REVIEW

File: lib/ecto_job/supervisor.ex [hunk 1]
```diff
@@ -33,15 +33,15 @@ defmodule EctoJob.Supervisor do
...
+  def start_link(config = %Config{repo: repo, schema: schema, max_demand: max_demand, poll_interval: poll_interval, base_expiry_seconds: base_expiry_seconds}) do
     supervisor_name = String.to_atom("#{schema}.Supervisor")
     notifier_name = String.to_atom("#{schema}.Notifier")
     producer_name = String.to_atom("#{schema}.Producer")
 
     children = [
       worker(Postgrex.Notifications, [repo.config() ++ [name: notifier_name]]),
       worker(Producer, [
-        [name: producer_name, repo: repo, schema: schema, notifier: notifier_name, poll_interval: poll_interval]
+        [name: producer_name, repo: repo, schema: schema, notifier: notifier_name, poll_interval: poll_interval, base_expiry_seconds: base_expiry_seconds]
```
Comment (mmartinson) [inline lib/ecto_job/supervisor.ex]: I think it'd be good to run the code formatter to address the line length on some of these. Figured it could be a different PR and do the whole codebase.

File: test/job_queue_test.exs [hunk 2]
```diff
@@ -64,7 +64,7 @@ defmodule EctoJob.JobQueueTest do
   end
 
   describe "JoqQueue.activate_scheduled_jobs" do
-    test "Updates scheduled job to active" do
+    test "Updates scheduled job to AVAILABLE" do
```
Comment (mmartinson) [inline test/job_queue_test.exs:69]: I found that when I was reading these tests to get acquainted with the codebase that the terminology was a touch confusing. I just edited for consistency.

File: README.md [hunk 3]
```diff
@@ -190,15 +196,23 @@ Scheduled jobs transition to "AVAILABLE" after the scheduled time has passed.
...
 Expired jobs in the "IN_PROGRESS" state with attempts >= MAX_ATTEMPTS move to a "FAILED" state.
 Failed jobs are kept in the database so that application developers can handle the failure.
+
+## Job Timeouts and Transactional Safety
+
+When performing long-running jobs or when configuring a low expiry timout, keep in mind that a job may be retried before it has finished and the retry has no proactive mechanism to cancel the running job.
+
+In the case that the initial job attempts to finish and commit a result, and the commit includes the preloaded multi passed as the first parameter to `perform/2`, the optimistic lock will fail the transaction.
+
+In the case where the job performs other side effects outside of the transaction such as calls to external APIs or additional database writes, these are suggested to implment other idempotency guarantees, as they will not be rolled back in a failed or duplicated job.
```
Comment (mmartinson) [inline README.md]: Let me know if this sounds correct and helpful

File: README.md [hunk 4]
```diff
@@ -190,15 +196,23 @@ Scheduled jobs transition to "AVAILABLE" after the scheduled time has passed.
...
 
-Once a consumer is given a job, it increments the attempt counter and updates the state to "IN_PROGRESS", with an expiry of 5 minutes.
-If the job is being retried, the expiry will be 5 minutes * the attempt counter.
+Once a consumer is given a job, it increments the attempt counter and updates the state to "IN_PROGRESS", with an initial expiry of the configurable as `base_expiry_seconds`, which defaults to 5 minutes.
+If the job is being retried, the expiry will be base expiry * the attempt counter.
 
-If successful, the consumer will delete the job from the queue.
-If unsuccessful, the job remains in the "IN_PROGRESS" state until it expires.
+If successful, the consumer can delete the job from the queue using the preloaded multi passed to the `perform/2` job handler.
+If an exception is raised in the worker or a successful processing attempt failes to successfully comitt the preloaded multi, the job is not deleted and remains in the "IN_PROGRESS" state until it expires.
```
Comment (mbuhot) [inline README.md]: failes -> fails

comitt -> commit

File: README.md (hunk 3 above)
Comment (mbuhot) [reply README.md]: This is a great description of how it works today. 👍 
I'm sure the mechanism could be improved. Eg, some way for a job to signal that it is still alive and the expiry should be extended.

Comment (mbuhot) [reply README.md]: implment -> implement

File: lib/ecto_job/config.ex [hunk 5]
```diff
@@ -22,6 +22,7 @@ defmodule EctoJob.Config do
     - `log`: (Default `true`) Enables logging using the standard Elixir `Logger` module
     - `log_level`: (Default `:info`) The level used to log messages, see [Logger](https://hexdocs.pm/logger/Logger.html#module-levels)
     - `poll_interval`: (Default `60_000`) Time in milliseconds between polling the `JobQueue` for scheduled jobs or jobs due to be retried
+    - `base_expiry_seconds`: (Default `300`) Time in seconds where a `RESERVED` or `IN_PROGRESS` job state is held before subsequent polls return a job to the `AVAILABLE` state for retry. The time will double for every retry until `max_attemps` is reached for a given job.
```
Comment (mbuhot) [inline lib/ecto_job/config.ex]: > The time will double for every retry

How about: "The time is extended by `base_expiry_seconds` for every retry" ?

File: lib/ecto_job/producer.ex [hunk 6]
```diff
@@ -24,16 +24,17 @@ defmodule EctoJob.Producer do
...
 
     @type t :: %__MODULE__{
             repo: EctoJob.Producer.repo(),
             schema: EctoJob.Producer.schema(),
             notifier: EctoJob.Producer.notifier(),
             demand: integer,
             clock: (() -> DateTime.t()),
-            poll_interval: non_neg_integer()
+            poll_interval: non_neg_integer(),
+            base_expiry_seconds: non_neg_integer()
```
Comment (mbuhot) [inline lib/ecto_job/producer.ex]: For someone unfamiliar with the code, I wonder if a name like `execution_timeout` would be more obvious?
I'm tempted to also configure `reservation_timeout` separately, since it is safer to set to a lower value.

File: lib/ecto_job/job_queue.ex [hunk 7]
```diff
@@ -309,12 +313,15 @@ defmodule EctoJob.JobQueue do
 
   @doc """
   Computes the expiry time for an `"IN_PROGRESS"` job based on the current time and attempt counter
+
+  Note that the integer value represents seconds to conform with unix timestamp values,
+  not milliseconds
   """
   @spec progress_expiry(DateTime.t(), integer) :: DateTime.t()
-  def progress_expiry(now, attempt) do
+  def progress_expiry(now, attempt, base_expiry_seconds \\ 300) do
```
Comment (mbuhot) [inline lib/ecto_job/job_queue.ex]: Since the 300s value is defaulted in `Config`, I think it is fine to remove the `\\ 300` here.

File: lib/ecto_job/job_queue.ex [hunk 8]
```diff
@@ -224,11 +224,11 @@ defmodule EctoJob.JobQueue do
   The batch size is determined by `demand`.
   returns `{count, updated_jobs}` tuple.
   """
-  @spec reserve_available_jobs(repo, schema, integer, DateTime.t()) :: {integer, [job]}
-  def reserve_available_jobs(repo, schema, demand, now = %DateTime{}) do
+  @spec reserve_available_jobs(repo, schema, integer, DateTime.t(), integer) :: {integer, [job]}
+  def reserve_available_jobs(repo, schema, demand, now = %DateTime{}, base_expiry \\ 300) do
```
Comment (mbuhot) [inline lib/ecto_job/job_queue.ex]: Since the 300s is defaulted in `Config` I think it is fine to remove the `\\ 300` here.

File: lib/ecto_job/job_queue.ex (hunk 7 above)
Comment (mmartinson) [reply lib/ecto_job/job_queue.ex]: Sure thing. I only added the default because this is a public function and it would be a breaking change if someone was calling it in the application code.

File: lib/ecto_job/job_queue.ex (hunk 8 above)
Comment (mmartinson) [reply lib/ecto_job/job_queue.ex]: As above, would be a breaking change. Would be fine to change it if you think that's ok in the given circumstances

File: lib/ecto_job/producer.ex (hunk 6 above)
Comment (mmartinson) [reply lib/ecto_job/producer.ex]: Those both sound like fine changes to me. I'll have a chance to edit in a couple days. I thought it might be helpful to specify the unit in the name, since this is the only case where the integer is seconds and not ms, but maybe mentioning in the docs is enough. Alternatively, I could change it to a ms value to normalize it.

File: README.md (hunk 4 above)
Comment (mmartinson) [reply README.md]: thanks. was finishing up this part quickly :smile:

File: lib/ecto_job/producer.ex (hunk 6 above)
Comment (mbuhot) [reply lib/ecto_job/producer.ex]: 👍 let's use ms

**Comment by mmartinson** [comment]
@mbuhot I went through and made the changes we were discussing. I have this set up to do some further testing with my app and I'll let you know when I can test that it's working as expected. 

One minor thing I noticed when changing the integer unit to ms is that we need to round back to seconds regardless if we're going to keep storing the time as a unix timestamp integer. I used floor div, and I'm not sure if/where it merits documenting further. I'm inclined to think that it's better this way from a perspective of understandability, and I doubt people are/will be concerned about tweaking the sub-second timeout values.

I agree, as you suggested, that it would be great to have a mechanism to notify a job that it has been timed out. I would prefer to not tackle it in this PR, but would be happy to chat further about what an API and implementation for it might look like.

Let me know if you have any further suggestions or changes you'd like to see.

File: README.md [hunk 9]
```diff
@@ -152,6 +152,18 @@ You can customize how often the table is polled for scheduled jobs.  The default
 config :ecto_job, :poll_interval, 15_000
 ```
 
+Control the time for which the job is reserved while waiting for a worker to pick it up, before the poller will make the job available again for dispatch by the producer.  The default is `60_000` ms.
+
+```
+config :ecto_job, :reservation_timeout, 15_000
+```
+
+Control the timeout for job execution before a job will be made available for retry. Begins when job is picked up by worker. Keep in mind, for jobs that are expected to retry quickly, any configured `execution_timout` will only retry a job as quickly as the `poll_interval`. An exception to this may be the first retry, which may not be bounded similarly if the initial job trigger dispatches the first attempt mid-interval. The default is `300_000` ms (5 mins).
```
Comment (mbuhot) [inline README.md]: I think we can leave out the special case of the first retry.

**Review by mbuhot — APPROVED** [review]
@mmartinson This is looking great, thanks again 👍

**Comment by mmartinson** [comment]
Cool, thanks. Initial testing of this has it working as expected. I'll be doing some more in depth testing in a prod environment this week if you want to wait to merge it, though if you want to go ahead if there is anything that comes up I'll be jumping in with a fix.
